llvm/lib/Transforms/Vectorize/VPlan.h

```diff
@@ -2296,29 +2304,34 @@ class VPReductionRecipe : public VPSingleDefRecipe {
 protected:
   VPReductionRecipe(const unsigned char SC, const RecurrenceDescriptor &R,
                     Instruction *I, ArrayRef<VPValue *> Operands,
-                    VPValue *CondOp, bool IsOrdered, DebugLoc DL)
-      : VPSingleDefRecipe(SC, Operands, I, DL), RdxDesc(R),
+                    VPValue *CondOp, bool IsOrdered)
+      : VPRecipeWithIRFlags(SC, Operands, *I), RdxDesc(R),
         IsOrdered(IsOrdered) {
     if (CondOp) {
```

Comment:
Should we set the RecurrenceDescriptor FMFs directly?

```suggestion
                    VPValue *CondOp, bool IsOrdered)
      : VPRecipeWithIRFlags(SC, Operands, R.getFastMathFlags(), I->getDebugLoc()), RdxDesc(R),
        IsOrdered(IsOrdered) {
    setUnderlyingValue(I);
    if (CondOp) {
```

I think this is closer to NFC. And that way we always have FMFs and you can remove the checks to `if (hasFastMathFlags())`.

Reply:
No, we cannot construct VPReductionRecipe without underlying value at least for now. 

1. In VPRecipeBase::cost(), `forceTargetInstructionCost` and `skipCostComputation` will check if the  recipe contains underlying instruction and override/skip the cost of the recipe. 

I think is better to set the flag directly after the #113903 landed and a better logic to override the cost from `forceTargetInstructionCost`.

Reply:
But this sets the underlying instruction on line 2311? Without this we end up with the non-FMF flag in llvm/test/Transforms/LoopVectorize/RISCV/vplan-vp-intrinsics-reduction.ll

Reply:
Oops. My bad! I didn't notice the `setUnderlyingValue()`.

But this will end up with some weird test cases.
The RecurrenceDescriptor default contains `fast` flags.  If we always set the FMF directly from recurrenceDescriptor without check if the underlyingInstruction is FPMathOperator or not, we will see `fast` on the integer reduction. 

Perhaps clear the flags carrying by the underlying instruction in the `VPReductionRecipe` and set the `FMF` if the underlying instruction is `FPMathOperator` is a better solution to avoid tests changes.

Reply:
That makes sense to me. 
Updated, thanks !